decoder-loop-example-parse-into-recogs-style-variables.rasp

```diff
@@ -0,0 +1,390 @@
+# Example building on example-word-type-dictionary-lookup.rasp to demo a basic decoder loop parsing out nouns and verbs as numbered variable in a logical form (somewhat ReCOGS style)
+# using Restricted Access Sequence Processing (RASP) (Weiss et al 2021)
+# which expresses the operations Transformer neural networks can perform
+# (and can be compiled to a neural network).
+
+# WARNING: This is just an example and should not be used for anything as is!
+
+# (some few nouns and verbs from ReCOGS vocab, not using more for easier reading)
+set example "nouns: girl boy cat dog box drink emma liam . verbs: ate painted drew loved . ignore: the a . the boy painted a girl ."
```

Comment:
Here I am starting by just labeling each word as either a noun, verb, or ignore in a single RASP pass after embedding the words in the input.

However I will attempt applying this to the more complicated grammar of possible COGS sentences (Kim and Linzen 2020; "COGS: A Compositional Generalization Challenge Based on Semantic Interpretation" , https://aclanthology.org/2020.emnlp-main.731 ).

IBM's CPG project (not their solution but utilities) gives the grammar for the sentences in  COGS from Kim and Linzen 2020 ("COGS: A Compositional Generalization Challenge Based on Semantic Interpretation" , https://aclanthology.org/2020.emnlp-main.731) at https://github.com/IBM/cpg/blob/c3626b4e03bfc681be2c2a5b23da0b48abe6f570/src/model/cogs_data.py#L523 as:
```
start: s1 | s2 | s3 | s4 | vp_internal
    s1: np vp_external
    s2: np vp_passive
    s3: np vp_passive_dat
    s4: np vp_external4
    vp_external: v_unerg | v_trans_omissible_p1 | vp_external1 | vp_external2 | vp_external3 | vp_external5 | vp_external6 | vp_external7
    vp_external1: v_unacc_p1 np
    vp_external2: v_trans_omissible_p2 np
    vp_external3: v_trans_not_omissible np
    vp_external4: v_inf_taking to v_inf
    vp_external5: v_cp_taking that start
    vp_external6: v_dat_p1 np pp_iobj
    vp_external7: v_dat_p2 np np
    vp_internal: np v_unacc_p2
    vp_passive: vp_passive1 | vp_passive2 | vp_passive3 | vp_passive4 | vp_passive5 | vp_passive6 | vp_passive7 | vp_passive8
    vp_passive1: was v_trans_not_omissible_pp_p1
    vp_passive2: was v_trans_not_omissible_pp_p2 by np
    vp_passive3: was v_trans_omissible_pp_p1
    vp_passive4: was v_trans_omissible_pp_p2 by np
    vp_passive5: was v_unacc_pp_p1
    vp_passive6: was v_unacc_pp_p2 by np
    vp_passive7: was v_dat_pp_p1 pp_iobj
    vp_passive8: was v_dat_pp_p2 pp_iobj by np
    vp_passive_dat: vp_passive_dat1 | vp_passive_dat2
    vp_passive_dat1: was v_dat_pp_p3 np
    vp_passive_dat2: was v_dat_pp_p4 np by np
    np: np_prop | np_det | np_pp
    np_prop: proper_noun
    np_det: det common_noun
    np_pp: np_det pp np
    pp_iobj: to np
    det: "the" | "a"
    pp: "on" | "in" | "beside"
    was: "was"
    by: "by"
    to: "to"
    that: "that"
    common_noun: "girl" | "boy" | "cat" | "dog" | "baby" | "child" | "teacher" | "frog" | "chicken" | "mouse" | "lion" | "monkey" | "bear" | "giraffe" | "horse" | "bird" | "duck" | "bunny" | "butterfly" | "penguin" | "student" | "professor" | "monster" | "hero" | "sailor" | "lawyer" | "customer" | "scientist" | "princess" | "president" | "cow" | "crocodile" | "goose" | "hen" | "deer" | "donkey" | "bee" | "fly" | "kitty" | "tiger" | "wolf" | "zebra" | "mother" | "father" | "patient" | "manager" | "director" | "king" | "queen" | "kid" | "fish" | "moose" | "pig" | "pony" | "puppy" | "sheep" | "squirrel" | "lamb" | "turkey" | "turtle" | "doctor" | "pupil" | "prince" | "driver" | "consumer" | "writer" | "farmer" | "friend" | "judge" | "visitor" | "guest" | "servant" | "chief" | "citizen" | "champion" | "prisoner" | "captain" | "soldier" | "passenger" | "tenant" | "politician" | "resident" | "buyer" | "spokesman" | "governor" | "guard" | "creature" | "coach" | "producer" | "researcher" | "guy" | "dealer" | "duke" | "tourist" | "landlord" | "human" | "host" | "priest" | "journalist" | "poet" | "hedgehog" | "shark" | "cockroach" | "cobra" | "hippo" | "cake" | "donut" | "cookie" | "box" | "rose" | "drink" | "raisin" | "melon" | "sandwich" | "strawberry" | "ball" | "balloon" | "bat" | "block" | "book" | "crayon" | "chalk" | "doll" | "game" | "glue" | "lollipop" | "hamburger" | "banana" | "biscuit" | "muffin" | "pancake" | "pizza" | "potato" | "pretzel" | "pumpkin" | "sweetcorn" | "yogurt" | "pickle" | "jigsaw" | "pen" | "pencil" | "present" | "toy" | "cracker" | "brush" | "radio" | "cloud" | "mandarin" | "hat" | "basket" | "plant" | "flower" | "chair" | "spoon" | "pillow" | "gumball" | "scarf" | "shoe" | "jacket" | "hammer" | "bucket" | "knife" | "cup" | "plate" | "towel" | "bottle" | "bowl" | "can" | "clock" | "jar" | "penny" | "purse" | "soap" | "toothbrush" | "watch" | "newspaper" | "fig" | "bag" | "wine" | "key" | "weapon" | "brain" | "tool" | "crown" | "ring" | "leaf" | "fruit" | "mirror" | "beer" | "shirt" | "guitar" | "chemical" | "seed" | "shell" | "brick" | "bell" | "coin" | "button" | "needle" | "molecule" | "crystal" | "flag" | "nail" | "bean" | "liver" | "table" | "stage" | "bed" | "chair" | "stool" | "road" | "tree" | "box" | "surface" | "seat" | "speaker" | "computer" | "rock" | "boat" | "cabinet" | "tv" | "plate" | "desk" | "bowl" | "bench" | "shelf" | "cloth" | "piano" | "bible" | "leaflet" | "sheet" | "cupboard" | "truck" | "tray" | "notebook" | "blanket" | "deck" | "coffin" | "log" | "ladder" | "barrel" | "rug" | "canvas" | "tiger" | "towel" | "throne" | "booklet" | "sock" | "corpse" | "sofa" | "keyboard" | "book" | "pillow" | "pad" | "train" | "couch" | "bike" | "pedestal" | "platter" | "paper" | "rack" | "board" | "panel" | "tripod" | "branch" | "machine" | "floor" | "napkin" | "cookie" | "block" | "cot" | "device" | "yacht" | "dog" | "mattress" | "ball" | "stand" | "stack" | "windowsill" | "counter" | "cushion" | "hanger" | "trampoline" | "gravel" | "cake" | "carpet" | "plaque" | "boulder" | "leaf" | "mound" | "bun" | "dish" | "cat" | "podium" | "tabletop" | "beach" | "bag" | "glacier" | "brick" | "crack" | "vessel" | "futon" | "turntable" | "rag" | "chessboard" | "house" | "room" | "car" | "garden" | "box" | "cup" | "glass" | "bag" | "vehicle" | "hole" | "cabinet" | "bottle" | "shoe" | "storage" | "cot" | "vessel" | "pot" | "pit" | "tin" | "can" | "cupboard" | "envelope" | "nest" | "bush" | "coffin" | "drawer" | "container" | "basin" | "tent" | "soup" | "well" | "barrel" | "bucket" | "cage" | "sink" | "cylinder" | "parcel" | "cart" | "sack" | "trunk" | "wardrobe" | "basket" | "bin" | "fridge" | "mug" | "jar" | "corner" | "pool" | "blender" | "closet" | "pile" | "van" | "trailer" | "saucepan" | "truck" | "taxi" | "haystack" | "dumpster" | "puddle" | "bathtub" | "pod" | "tub" | "trap" | "bun" | "microwave" | "bookstore" | "package" | "cafe" | "train" | "castle" | "bunker" | "vase" | "backpack" | "tube" | "hammock" | "stadium" | "backyard" | "swamp" | "monastery" | "refrigerator" | "palace" | "cubicle" | "crib" | "condo" | "tower" | "crate" | "dungeon" | "teapot" | "tomb" | "casket" | "jeep" | "shoebox" | "wagon" | "bakery" | "fishbowl" | "kennel" | "china" | "spaceship" | "penthouse" | "pyramid" | "table" | "stage" | "bed" | "chair" | "book" | "road" | "tree" | "machine" | "house" | "seat" | "speaker" | "computer" | "rock" | "car" | "box" | "cup" | "glass" | "bag" | "flower" | "boat" | "vehicle" | "key" | "painting" | "cabinet" | "tv" | "bottle" | "cat" | "desk" | "shoe" | "mirror" | "clock" | "bench" | "bike" | "lamp" | "lion" | "piano" | "crystal" | "toy" | "duck" | "sword" | "sculpture" | "rod" | "truck" | "basket" | "bear" | "nest" | "sphere" | "bush" | "surgeon" | "poster" | "throne" | "giant" | "trophy" | "hedge" | "log" | "tent" | "ladder" | "helicopter" | "barrel" | "yacht" | "statue" | "bucket" | "skull" | "beast" | "lemon" | "whale" | "cage" | "gardner" | "fox" | "sink" | "trainee" | "dragon" | "cylinder" | "monk" | "bat" | "headmaster" | "philosopher" | "foreigner" | "worm" | "chemist" | "corpse" | "wolf" | "torch" | "sailor" | "valve" | "hammer" | "doll" | "genius" | "baron" | "murderer" | "bicycle" | "keyboard" | "stool" | "pepper" | "warrior" | "pillar" | "monkey" | "cassette" | "broker" | "bin"
    proper_noun: "emma" | "liam" | "olivia" | "noah" | "ava" | "william" | "isabella" | "james" | "sophia" | "oliver" | "charlotte" | "benjamin" | "mia" | "elijah" | "amelia" | "lucas" | "harper" | "mason" | "evelyn" | "logan" | "abigail" | "alexander" | "emily" | "ethan" | "elizabeth" | "jacob" | "mila" | "michael" | "ella" | "daniel" | "avery" | "henry" | "sofia" | "jackson" | "camila" | "sebastian" | "aria" | "aiden" | "scarlett" | "matthew" | "victoria" | "samuel" | "madison" | "david" | "luna" | "joseph" | "grace" | "carter" | "chloe" | "owen" | "penelope" | "wyatt" | "layla" | "john" | "riley" | "jack" | "zoey" | "luke" | "nora" | "jayden" | "lily" | "dylan" | "eleanor" | "grayson" | "hannah" | "levi" | "lillian" | "isaac" | "addison" | "gabriel" | "aubrey" | "julian" | "ellie" | "mateo" | "stella" | "anthony" | "natalie" | "jaxon" | "zoe" | "lincoln" | "leah" | "joshua" | "hazel" | "christopher" | "violet" | "andrew" | "aurora" | "theodore" | "savannah" | "caleb" | "audrey" | "ryan" | "brooklyn" | "asher" | "bella" | "nathan" | "claire" | "thomas" | "skylar" | "leo" | "lina" | "paula" | "charlie"
    v_trans_omissible_p1: "ate" | "painted" | "drew" | "cleaned" | "cooked" | "dusted" | "hunted" | "nursed" | "sketched" | "juggled" | "called" | "heard" | "packed" | "saw" | "noticed" | "studied" | "examined" | "observed" | "knew" | "investigated" | "baked"
    v_trans_omissible_p2: "ate" | "painted" | "drew" | "cleaned" | "cooked" | "dusted" | "hunted" | "nursed" | "sketched" | "juggled" | "called" | "heard" | "packed" | "saw" | "noticed" | "studied" | "examined" | "observed" | "knew" | "investigated" | "baked"
    v_trans_omissible_pp_p1: "eaten" | "painted" | "drawn" | "cleaned" | "cooked" | "dusted" | "hunted" | "nursed" | "sketched" | "juggled" | "called" | "heard" | "packed" | "seen" | "noticed" | "studied" | "examined" | "observed" | "known" | "investigated"
    v_trans_omissible_pp_p2: "eaten" | "painted" | "drawn" | "cleaned" | "cooked" | "dusted" | "hunted" | "nursed" | "sketched" | "juggled" | "called" | "heard" | "packed" | "seen" | "noticed" | "studied" | "examined" | "observed" | "known" | "investigated"
    v_trans_not_omissible: "liked" | "helped" | "found" | "loved" | "poked" | "admired" | "adored" | "appreciated" | "missed" | "respected" | "threw" | "tolerated" | "valued" | "worshipped" | "discovered" | "held" | "stabbed" | "touched" | "pierced" | "tossed"
    v_trans_not_omissible_pp_p1: "liked" | "helped" | "found" | "loved" | "poked" | "admired" | "adored" | "appreciated" | "missed" | "respected" | "thrown" | "tolerated" | "valued" | "worshipped" | "discovered" | "held" | "stabbed" | "touched" | "pierced" | "tossed"
    v_trans_not_omissible_pp_p2: "liked" | "helped" | "found" | "loved" | "poked" | "admired" | "adored" | "appreciated" | "missed" | "respected" | "thrown" | "tolerated" | "valued" | "worshipped" | "discovered" | "held" | "stabbed" | "touched" | "pierced" | "tossed"
    v_cp_taking: "liked" | "hoped" | "said" | "noticed" | "believed" | "confessed" | "declared" | "proved" | "thought" | "admired" | "appreciated" | "respected" | "supported" | "tolerated" | "valued" | "wished" | "dreamed" | "expected" | "imagined" | "meant"
    v_inf_taking: "wanted" | "preferred" | "needed" | "intended" | "tried" | "attempted" | "planned" | "expected" | "hoped" | "wished" | "craved" | "liked" | "hated" | "loved" | "enjoyed" | "dreamed" | "meant" | "longed" | "yearned" | "itched"
    v_unacc_p1: "rolled" | "froze" | "burned" | "shortened" | "floated" | "grew" | "slid" | "broke" | "crumpled" | "split" | "changed" | "snapped" | "disintegrated" | "collapsed" | "decomposed" | "doubled" | "improved" | "inflated" | "enlarged" | "reddened" | "shattered" | "blessed" | "squeezed"
    v_unacc_p2: "rolled" | "froze" | "burned" | "shortened" | "floated" | "grew" | "slid" | "broke" | "crumpled" | "split" | "changed" | "snapped" | "disintegrated" | "collapsed" | "decomposed" | "doubled" | "improved" | "inflated" | "enlarged" | "reddened" | "shattered" | "blessed" | "squeezed"
    v_unacc_pp_p1: "rolled" | "frozen" | "burned" | "shortened" | "floated" | "grown" | "slid" | "broken" | "crumpled" | "split" | "changed" | "snapped" | "disintegrated" | "collapsed" | "decomposed" | "doubled" | "improved" | "inflated" | "enlarged" | "reddened" | "shattered" | "blessed" | "squeezed"
    v_unacc_pp_p2: "rolled" | "frozen" | "burned" | "shortened" | "floated" | "grown" | "slid" | "broken" | "crumpled" | "split" | "changed" | "snapped" | "disintegrated" | "collapsed" | "decomposed" | "doubled" | "improved" | "inflated" | "enlarged" | "reddened" | "shattered" | "blessed" | "squeezed"
    v_unerg: "slept" | "smiled" | "laughed" | "sneezed" | "cried" | "talked" | "danced" | "jogged" | "walked" | "ran" | "napped" | "snoozed" | "screamed" | "stuttered" | "frowned" | "giggled" | "scoffed" | "snored" | "smirked" | "gasped"
    v_inf: "walk" | "run" | "sleep" | "sneeze" | "nap" | "eat" | "read" | "cook" | "hunt" | "paint" | "talk" | "dance" | "giggle" | "jog" | "smirk" | "call" | "sketch" | "dust" | "clean" | "investigate" | "crawl"
    v_dat_p1: "gave" | "lended" | "sold" | "offered" | "fed" | "passed" | "sent" | "rented" | "served" | "awarded" | "brought" | "handed" | "forwarded" | "promised" | "mailed" | "loaned" | "posted" | "returned" | "slipped" | "wired" | "teleported" | "shipped"
    v_dat_p2: "gave" | "lended" | "sold" | "offered" | "fed" | "passed" | "sent" | "rented" | "served" | "awarded" | "brought" | "handed" | "forwarded" | "promised" | "mailed" | "loaned" | "posted" | "returned" | "slipped" | "wired" | "teleported" | "shipped"
    v_dat_pp_p1: "given" | "lended" | "sold" | "offered" | "fed" | "passed" | "sent" | "rented" | "served" | "awarded" | "brought" | "handed" | "forwarded" | "promised" | "mailed" | "loaned" | "posted" | "returned" | "slipped" | "wired"
    v_dat_pp_p2: "given" | "lended" | "sold" | "offered" | "fed" | "passed" | "sent" | "rented" | "served" | "awarded" | "brought" | "handed" | "forwarded" | "promised" | "mailed" | "loaned" | "posted" | "returned" | "slipped" | "wired"
    v_dat_pp_p3: "given" | "lended" | "sold" | "offered" | "fed" | "passed" | "sent" | "rented" | "served" | "awarded" | "brought" | "handed" | "forwarded" | "promised" | "mailed" | "loaned" | "posted" | "returned" | "slipped" | "wired"
    v_dat_pp_p4: "given" | "lended" | "sold" | "offered" | "fed" | "passed" | "sent" | "rented" | "served" | "awarded" | "brought" | "handed" | "forwarded" | "promised" | "mailed" | "loaned" | "posted" | "returned" | "slipped" | "wired"
```

Note that RASP does NOT stand for Robust Accurate Statistical Parsing in this project, it is Restricted Access Sequence Processing from Weiss et al 2021 ( "Thinking Like Transformers" , https://arxiv.org/abs/2106.06981 ), and I have not observed anyone building parsers in this RASP.

However, we can see using approach above one way in Restricted Access Sequence Processing (RASP) to tag the words as 
```
det: "the" | "a"
pp: "on" | "in" | "beside"
was: "was"
by: "by"
to: "to"
that: "that"
common_noun: "girl" | "boy" | "cat" | "dog"| ...
proper_noun: "emma" | "liam" | "olivia" | ...
v_trans_omissible: "ate" | "painted" | "drew" | ...
v_trans_not_omissible: "liked" | "helped" | "found" | ...
v_cp_taking: "liked" | "hoped" | "said" | ...
v_inf_taking: "wanted" | "preferred" | "needed" | ...
[and so on]
```
Noting of course that in some cases we will need to tag the same words with multiple tags.
We can have sequences which support different subsets of tags as well as a sequence just representing each tag independently.

Analyzing recursive cases like recursive prepositional clauses suggests it might be the case that the number of layers in our Transformer will depend on the recursion depth we are supporting (if we don't use a scratchpad and process the sequence iteratively there), as each time we tag a pattern in our sequence, the RASP operation increases the Transformer depth.

Specifically, the recursive preposition case I am referring to is given:
```
np: np_prop | np_det | np_pp
np_prop: proper_noun
np_det: det common_noun
np_pp: np_det pp np
det: "the" | "a"
pp: "on" | "in" | "beside"
```
One sees that `np_pp` can expand to `np_det pp np` where `np` can then expand to `np_pp` again,
like `np_pp -> np_det pp np -> np_det pp (np_det pp np) -> np_det pp np_det pp np` some potentially unbounded number of times. 

However the Transformer in a fixed depth may (also or instead of handling it recursively) memorize a shortcut mapping for a particular depth like `np_det pp np_det pp np_det pp [..repeated some specific N times..] np_prop` and do that mapping with a fixed depth, and I can probably program this in RASP and compare. We shall see!

Csordas et al 2022 ( https://arxiv.org/abs/2110.07732 ) predict that a Transformer doing compositional operations without scratchpad will have required depth proportional to the depth of the computational graph of the computation it is programming (depth of parse tree here) and that for repeated compositional steps the weights may be tied for the layers for those steps. 

When doing this by hand the obvious way, it seems the earliest layers will be tagging `det, pp, was, by, to, that, common_noun, proper_noun, v_trans_omissible, v_trans_omissible_pp, v_trans_not_omissible, v_trans_not_omissible_pp, v_cp_taking, v_inf_taking, v_unacc, v_unacc_pp, v_unerg, v_inf, v_dat, v_dat_pp`

but that we will end up with some deeper stack of layers after that which may have similar weights repeated (but not similar to the early layers recognizing `det, pp, ...`) implementing repeated operations to process higher order grammar and potentially recursive items like `np: np_prop | np_det | np_pp` and `np_pp: np_det pp np`.

Reply:
Adding detail to the previous comment.
For the example sentence `the boy painted a girl`, we would use the following lines from the lookup table:

(obtained via IBM's CPG project (not their solution but utilities) which gives the grammar for the sentences in COGS from Kim and Linzen 2020 ("COGS: A Compositional Generalization Challenge Based on Semantic Interpretation" , https://aclanthology.org/2020.emnlp-main.731) at https://github.com/IBM/cpg/blob/c3626b4e03bfc681be2c2a5b23da0b48abe6f570/src/model/cogs_data.py#L523 )

````
start: s1 | s2 | s3 | s4 | vp_internal
s1: np vp_external
...
vp_external: v_unerg | v_trans_omissible_p1 | vp_external1 | vp_external2 | vp_external3 | vp_external5 | 
...
vp_external2: v_trans_omissible_p2 np
...
np: np_prop | np_det | np_pp
...
np_det: det common_noun
...
det: "the" | "a"
...
common_noun: "girl" | "boy" | "cat" | "dog"| ...
...
v_trans_omissible: "ate" | "painted" | "drew" | ...
```
and in the first step instead of deriving a sequence with words replaced by tags noun/verb/ignore which is used to reorder them and list them out as in this PR, we would derive a sequence from `the boy painted a girl` where these words are replaced with: 

`det common_noun v_trans_omissible det common_noun`

In the next step, we can further lookup `det common_noun` and replace it with `np_det` to get:

`np_det v_trans_omissible np_det`

which can be simplified in additional passes via `np v_trans_omissible_p2 np -> np vp_external2 -> np vp_external -> s1`.

To generate the logical form we can use the templates for each part (simplified COGS where semantically irrelevant `x_`, ";", `AND` are removed, but not full ReCOGS simplified form).
For np_det with det=="the", we use `* noun ( idx )`  , 
for np_det with det=="a" we use `noun ( idx )`,
for `v_trans_omissible_p2` (p2 means np after),
which is  `verb . agent ( verb_idx , agent_idx ) verb . theme ( verb_idx , theme_idx )`

which gives (labeling each noun with its 1-based index in the original text per simplified COGS):
`* boy ( 2 ) painted . agent ( 3 , 1 ) painted . theme ( 3 , 5 ) girl ( 5 )` in simplified* COGS format (`painted` would also be mapped to `paint` separately).

* only simplification is removing `x_`, ";", `AND` filler text which distracts from the example.

As discussed in the code comments, I will also be translating these to ReCOGS form which is a bit different, see Zhengxuan et al 2024 ("ReCOGS: How Incidental Details of a Logical Form Overshadow an Evaluation of Semantic Interpretation", https://arxiv.org/abs/2303.13716 ).